### PR TITLE
Port class format simplifications to MATLAB

### LINKED ISSUE #1543: Adopt new Slice class format and preservation approach
In "new Slice" (https://docs.icerpc.dev/slice1):
 - exceptions are always encoded in the sliced format
 - classes encoded in Sliced format are always "fully preserved" (there is no attribute to preserve slices)
 - exceptions are not preserved - unknown slices are lost during decoding

I propose we do the same in Ice for Ice 3.8 and cleanup Slice files like:
https://github.com/zeroc-ice/ice/blob/ce1649d310904937a3e70a0ac79dc0bcb3a7433a/slice/Glacier2/Session.ice#L37

In this file, we mark operations `format:slice` ostensibly to encode the one exception (CannotCreateSessionException) in the sliced format even though we don't define anywhere a derived exception.

### LINKED ISSUE #1945: MATLAB slicing test failures
```
*** [8/10] Running matlab/Ice/slicing/exceptions tests ***
[ running client/server test - 03/15/24 16:28:54 ]
- Config: tcp
("/usr/bin/python3"  /home/runner/work/ice/ice/python/test/TestHelper.py Server --Ice.Default.Host=127.0.0.1 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.IPv6=0 --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 env={'LD_LIBRARY_PATH': '/home/runner/work/ice/ice/cpp/lib/x86_64-linux-gnu', 'PYTHONPATH': '/home/runner/work/ice/ice/python/python:/home/runner/work/ice/ice/python/test/Ice/slicing/exceptions'})
(/usr/local/bin/run-matlab-command "cd '/home/runner/work/ice/ice/matlab/test/lib', runTest /home/runner/work/ice/ice/matlab/test/Ice/slicing/exceptions /home/runner/work/ice/ice/matlab/lib/x86_64-linux-gnu --Ice.Default.Host=127.0.0.1 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.IPv6=0" env={'LD_LIBRARY_PATH': '/home/runner/work/ice/ice/cpp/lib/x86_64-linux-gnu'})
testing stringToProxy... ok
testing checked cast... ok
base... ok
base (AMI)... ok
slicing of unknown derived... ok
slicing of unknown derived (AMI)... ok
non-slicing of known derived as base... ok
non-slicing of known derived as base (AMI)... ok
non-slicing of known derived as derived... ok
non-slicing of known derived as derived (AMI)... ok
slicing of unknown intermediate as base... ok
slicing of unknown intermediate as base (AMI)... ok
slicing of known intermediate as base... ok
slicing of known intermediate as base (AMI)... ok
slicing of known most derived as base... ok
slicing of known most derived as base (AMI)... ok
non-slicing of known intermediate as intermediate... ok
non-slicing of known intermediate as intermediate (AMI)... ok
non-slicing of known most derived as intermediate... ok
non-slicing of known most derived as intermediate (AMI)... ok
non-slicing of known most derived as most derived... ok
non-slicing of known most derived as most derived (AMI)... ok
slicing of unknown most derived, known intermediate as base... ok
slicing of unknown most derived, known intermediate as base (AMI)... ok
slicing of unknown most derived, known intermediate as intermediate... ok
slicing of unknown most derived, known intermediate as intermediate (AMI)... ok
slicing of unknown most derived, unknown intermediate thrown as base... ok
slicing of unknown most derived, unknown intermediate thrown as base (AMI)... ok
unknown most derived in compact format... ok
preserved exceptions... Error using Test.TestIntfPrx/unknownPreservedAsBase
::Ice::OperationNotExistException

Error in AllTests.allTests (line 493)
                test.unknownPreservedAsBase();

Error in Client (line 14)
    test = AllTests.allTests(helper);

Error in runTest (line 28)
        Client(varargin);
exit status 1
Traceback (most recent call last):
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2050, in run
    self.runWithDriver(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2081, in runWithDriver
    current.driver.runClientServerTestCase(current)
  File "/home/runner/work/ice/ice/scripts/LocalDriver.py", line 680, in runClientServerTestCase
    self.runner.runClientSide(client, current, host)
  File "/home/runner/work/ice/ice/scripts/LocalDriver.py", line 207, in runClientSide
    testcase._runClientSide(current, host)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2027, in _runClientSide
    self.runClientSide(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 1913, in runClientSide
    self._runClient(current, client)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2039, in _runClient
    client.run(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 1441, in run
    process.waitSuccess(exitstatus=exitstatus, timeout=30)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 691, in waitSuccess
    self.testExitStatus(exitstatus)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 735, in testExitStatus
    test(self.exitstatus, exitstatus)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 717, in test
    raise RuntimeError(
RuntimeError: unexpected exit status: expected: 0, got 1
```

```
*** [9/10] Running matlab/Ice/slicing/objects tests ***
[ running client/server test - 03/15/24 16:29:03 ]
- Config: tcp
("/usr/bin/python3"  /home/runner/work/ice/ice/python/test/TestHelper.py Server --Ice.Default.Host=127.0.0.1 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.IPv6=0 --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 env={'LD_LIBRARY_PATH': '/home/runner/work/ice/ice/cpp/lib/x86_64-linux-gnu', 'PYTHONPATH': '/home/runner/work/ice/ice/python/python:/home/runner/work/ice/ice/python/test/Ice/slicing/objects'})
(/usr/local/bin/run-matlab-command "cd '/home/runner/work/ice/ice/matlab/test/lib', runTest /home/runner/work/ice/ice/matlab/test/Ice/slicing/objects /home/runner/work/ice/ice/matlab/lib/x86_64-linux-gnu --Ice.Default.Host=127.0.0.1 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.IPv6=0" env={'LD_LIBRARY_PATH': '/home/runner/work/ice/ice/cpp/lib/x86_64-linux-gnu'})
testing stringToProxy... ok
testing checked cast... ok
base as Object... ok
base as Object (AMI)... ok
base as Base... ok
base as Base (AMI)... ok
base with known derived as base... ok
base with known derived as base (AMI)... ok
base with known derived as known derived... ok
base with known derived as known derived (AMI)... ok
base with unknown derived as base... ok
base with unknown derived as base (AMI)... ok
unknown with Object as Object... ok
unknown with Object as Object (AMI)... ok
one-element cycle... ok
one-element cycle (AMI)... ok
two-element cycle... ok
two-element cycle (AMI)... ok
known derived pointer slicing as base... ok
known derived pointer slicing as base (AMI)... ok
known derived pointer slicing as derived... ok
known derived pointer slicing as derived (AMI)... ok
unknown derived pointer slicing as base... ok
unknown derived pointer slicing as base (AMI)... ok
param ptr slicing with known first... ok
param ptr slicing with known first (AMI)... ok
param ptr slicing with unknown first... ok
param ptr slicing with unknown first (AMI)... ok
return value identity with known first... ok
return value identity with known first (AMI)... ok
return value identity with unknown first... ok
return value identity with unknown first (AMI)... ok
return value identity for input params known first... Error using assert
Assertion failed.

Error in AllTests.allTests (line 484)
            assert(strcmp(b2.ice_id(), '::Test::B'));  % Sliced by server

Error in Client (line 14)
    test = AllTests.allTests(helper);

Error in runTest (line 28)
        Client(varargin);
exit status 1
Traceback (most recent call last):
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2050, in run
    self.runWithDriver(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2081, in runWithDriver
    current.driver.runClientServerTestCase(current)
  File "/home/runner/work/ice/ice/scripts/LocalDriver.py", line 680, in runClientServerTestCase
    self.runner.runClientSide(client, current, host)
  File "/home/runner/work/ice/ice/scripts/LocalDriver.py", line 207, in runClientSide
    testcase._runClientSide(current, host)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2027, in _runClientSide
    self.runClientSide(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 1913, in runClientSide
    self._runClient(current, client)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 2039, in _runClient
    client.run(current)
  File "/home/runner/work/ice/ice/scripts/Util.py", line 1441, in run
    process.waitSuccess(exitstatus=exitstatus, timeout=30)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 691, in waitSuccess
    self.testExitStatus(exitstatus)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 735, in testExitStatus
    test(self.exitstatus, exitstatus)
  File "/home/runner/work/ice/ice/scripts/Expect.py", line 717, in test
    raise RuntimeError(
RuntimeError: unexpected exit status: expected: 0, got 1
```

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR simplifies class (exception) unmarshaling in MATLAB:
 - classes are always preserved
 - exceptions are never preserved

It's somewhat simpler than in other mappings since we never marshal exceptions in MATLAB.

Fixes #1543 
Fixes #1945

### CODE REVIEW

File: matlab/test/Ice/slicing/objects/AllTests.m [hunk 1]
```diff
@@ -484,9 +481,13 @@
...
             assert(strcmp(b2.sb, 'D3.sb'));
-            assert(strcmp(b2.ice_id(), '::Test::B'));  % Sliced by server
-            assert(b2.pb == b1);
-            assert(~isa(b2, 'Test.D3'));
+            if proxy.ice_getEncodingVersion() == Ice.EncodingVersion(1, 0)
+                assert(~isa(b2, 'Test.D3'));
+            else
+                assert(isa(b2, 'Test.D3'));
+                assert(strcmp(b2.ice_id(), '::Test::D3'));  % preserved
+                assert(b2.pb == b1);
```
Comment (pepone) [inline matlab/test/Ice/slicing/objects/AllTests.m:489]: Should we put this assert outside the else block? `pb` is not Sliced for 1.0 encoding as it is part of the base class B.

File: matlab/test/Ice/slicing/objects/AllTests.m [hunk 2]
```diff
@@ -521,9 +522,13 @@
...
```
Comment (pepone) [inline matlab/test/Ice/slicing/objects/AllTests.m:530]: dito `assert(b2.pb == b1);` should work for both encodings.

**Review by pepone — APPROVED** [review]

File: matlab/test/Ice/slicing/objects/AllTests.m (hunk 1 above)
Comment (bernardnormier) [reply matlab/test/Ice/slicing/objects/AllTests.m:489]: Fixed

File: cpp/src/slice2matlab/Main.cpp [hunk 3]
```diff
@@ -1854,50 +1842,29 @@ CodeVisitor::visitClassDefStart(const ClassDefPtr& p)
         }
     }
 
-    if ((preserved && !basePreserved) || !convertMembers.empty())
+    if (!convertMembers.empty())
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2matlab/Main.cpp:1845]: It's unrelated to this PR.

**Review by externl — APPROVED** [review]

**Pull request merged**